File: libraries/ary/aryShift.c
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "ary_err.h"
#include "dat_par.h"

void aryShift( int nshift, const hdsdim *shift, Ary *ary, int *status ) {
/*
*+
*  Name:
*     aryShift

*  Purpose:
*     Apply pixel-index shifts to an array.

*  Synopsis:
*     void aryShift( int nshift, const hdsdim *shift, Ary *ary, int *status )

*  Description:
*     This function applies pixel-index shifts to an array. An integer
*     shift is applied to each dimension so that its pixel-index
*     bounds, and the indices of each pixel, change by the amount of
*     shift applied to the corresponding dimension. The array's pixels
*     retain their values and none are lost.

*  Parameters:
*     nshift
*        Number of dimensions to which shifts are to be applied (i.e. the
*        length of array "shift"). This must not exceed the number of array
*        dimensions. If fewer shifts are supplied than there are dimensions
*        in the array, then the extra dimensions will not be shifted.
*     shift
*        An array holding the pixel-index shifts to be applied to each
*        dimension.
*     ary
*        Array identifier.
*     status
*        The global status.

*  Notes:
*     -  Pixel-index shifts applied to a base array will affect the
*     appearance of that array as seen by all base-array identifiers
*     associated with it. However, array sections derived from that
*     base array will remain unchanged (as regards both pixel-indices
*     and data content).
*     -  Pixel-index shifts cannot be applied to a base array while any
*     part of it is mapped for access (i.e. even through another
*     identifier).
*     -  Pixel-index shifts applied to an array section only affect
*     that section itself, and have no effect on other array
*     identifiers.
*     -  Pixel-index shifts cannot be applied to an array section while
*     it is mapped for access through the identifier supplied to this
*     routine.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   AryACB *acb;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* If the number of shifts specified is not positive, then report an error. */
   if( nshift <= 0 ){
      *status = ARY__SFTIN;
      msgSeti( "BADNSFT", nshift );
      errRep( " ", "Invalid number of shifts (^BADNSFT) specified (possible "
              "programming error).", status );

/* Import the array identifier. */
   } else {
      acb = (AryACB *) ary1Impid( ary, 1, 0, 1, status );
      if( *status == SAI__OK ){

/* Check that the number of shifts does not exceed the number of array
   dimensions and report an error if it does. */
         if( nshift > acb->ndim ){
            *status = ARY__SFTIN;
            msgSeti( "BADNSFT", nshift );
            msgSeti( "NDIM", acb->ndim );
            errRep( " ", "Number of shifts specified (^BADNSFT) exceeds the "
                    "number of array dimensions (^NDIM) (possible "
                    "programming error).", status );

/* Check that SHIFT access to the array is available. */
         } else {
            ARY__DCB_LOCK_MUTEX;

            ary1Chacc( acb, "SHIFT", status );

/* Apply the shifts to the array's ACB entry. */
            ary1Sft( nshift, shift, acb, status );

            ARY__DCB_UNLOCK_MUTEX;
         }
      }
   }

/* If an error occurred, then report context information and call the error
   tracing routine. */
   if( *status != SAI__OK ){
      errRep( " ", "aryShift: Error applying pixel-index shifts to an array.",
              status );
      ary1Trace( "aryShift", status );
   }

}

File: libraries/fary/ary_shift.f
      SUBROUTINE ARY_SHIFT( NSHIFT, SHIFT, IARY, STATUS )
*+
*  Name:
*     ARY_SHIFT

*  Purpose:
*     Apply pixel-index shifts to an array.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY_SHIFT( NSHIFT, SHIFT, IARY, STATUS )

*  Description:
*     The routine applies pixel-index shifts to an array. An integer
*     shift is applied to each dimension so that its pixel-index
*     bounds, and the indices of each pixel, change by the amount of
*     shift applied to the corresponding dimension. The array's pixels
*     retain their values and none are lost.

*  Arguments:
*     NSHIFT = INTEGER (Given)
*        Number of dimensions to which shifts are to be applied. This
*        must not exceed the number of array dimensions. If fewer
*        shifts are supplied than there are dimensions in the array,
*        then the extra dimensions will not be shifted.
*     SHIFT( NSHIFT ) = INTEGER (Given)
*        The pixel-index shifts to be applied to each dimension.
*     IARY = INTEGER (Given)
*        Array identifier.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     -  Pixel-index shifts applied to a base array will affect the
*     appearance of that array as seen by all base-array identifiers
*     associated with it. However, array sections derived from that
*     base array will remain unchanged (as regards both pixel-indices
*     and data content).
*     -  Pixel-index shifts cannot be applied to a base array while any
*     part of it is mapped for access (i.e. even through another
*     identifier).
*     -  Pixel-index shifts applied to an array section only affect
*     that section itself, and have no effect on other array
*     identifiers.
*     -  Pixel-index shifts cannot be applied to an array section while
*     it is mapped for access through the identifier supplied to this
*     routine.

*  Algorithm:
*     -  Check that the number of pixel shifts specified is positive and
*     report an error if it is not.
*     -  Import the array identifier.
*     -  Check that the number of pixel shifts specified does not exceed
*     the number of array dimensions and report an error if it does.
*     -  Apply the shifts to the array.
*     -  If an error occurred, then report context information.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     8-AUG-1989 (RFWS):
*        Original version.
*     15-SEP-1989 (RFWS):
*        Added check that SHIFT access to the array is available.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Global Variables:
      INCLUDE 'ARY_ACB'          ! ARY_ Access Control Block
*        ACB_NDIM( ARY__MXACB ) = INTEGER (Read)
*           Number of array dimensions.

*  Arguments Given:
      INTEGER NSHIFT
      INTEGER SHIFT( * )
      INTEGER IARY

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER IACB               ! Index to array entry in the ACB

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  If the number of shifts specified is not positive, then report an
*  error.
      IF ( NSHIFT .LE. 0 ) THEN
         STATUS = ARY__SFTIN
         CALL MSG_SETI( 'BADNSFT', NSHIFT )
         CALL ERR_REP( 'ARY_SHIFT_NSLO',
     :   'Invalid number of shifts (^BADNSFT) specified (possible ' //
     :   'programming error).', STATUS )

*  Import the array identifier.
      ELSE
         CALL ARY1_IMPID( IARY, IACB, STATUS )
         IF ( STATUS .EQ. SAI__OK ) THEN

*  Check that the number of shifts does not exceed the number of array
*  dimensions and report an error if it does.
            IF ( NSHIFT .GT. ACB_NDIM( IACB ) ) THEN
               STATUS = ARY__SFTIN
               CALL MSG_SETI( 'BADNSFT', NSHIFT )
               CALL MSG_SETI( 'NDIM', ACB_NDIM( IACB ) )
               CALL ERR_REP( 'ARY_SHIFT_NSHI',
     :         'Number of shifts specified (^BADNSFT) exceeds the ' //
     :         'number of array dimensions (^NDIM) (possible ' //
     :         'programming error).', STATUS )

*  Check that SHIFT access to the array is available.
            ELSE
               CALL ARY1_CHACC( IACB, 'SHIFT', STATUS )

*  Apply the shifts to the array's ACB entry.
               CALL ARY1_SFT( NSHIFT, SHIFT, IACB, STATUS )
            END IF
         END IF
      END IF
       
*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'ARY_SHIFT_ERR',
     :   'ARY_SHIFT: Error applying pixel-index shifts to an array.',
     :   STATUS )
         CALL ARY1_TRACE( 'ARY_SHIFT', STATUS )
      END IF

      END
